Add support for upcoming invoices w/ invoice items

lskillen opened the pull request:
## Description

Adds support for Stripe's API for previewing upcoming invoices, and also adds support for subscription-based invoice items - This was a rather tricky one to implement given the ephemeral nature of the upcoming invoices.  They aren't persisted to the database and their internals differ from the expected "norm" (i.e. they don't have a stripe ID).

An example of usage (as we have it in our application):

```
invoice = djstripe_models.Invoice.upcoming(
    customer=customer, subscription=sub, subscription_plan=plan)
```

An additional `plan` property accessor has been added to invoices - As detailed in the code comments, this was necessary for us to provide a consistent view for the invoice history.  Instead of using the plan from the invoice subscription, it is now possible to use the plan for the from available subscription invoice item, and this access is encapsulated in the property.

The maximum length for stripe fields was also raised to 255, in accordance with the recommendations in the [Stripe API Upgrades documentation](https://stripe.com/docs/upgrades).  The main reason though was to make room for the longer composite key IDs used for subscription-based invoice items (necessary in order to ensure these remain unique per invoice).  As per the `plan` change above, this also helps to provide a consistent view for the invoice history.

One odd thing to note is the use of the `QuerySetMock` for `UpcomingInvoice`.  The reasoning behind this was to ensure that `invoice.invoiceitems()` for upcoming invoices still returned a queryset-like object that supports non-mutation operations such as iteration.  I made a note in the documentation not to expect it to do anything else, so it's a bit like Stripe's own docs for upcoming invoices, they are not mutable.  The main downside is the added dependency, so apologies for that!

Finally, I guess that the changes for `_get_or_create_from_stripe_object` might be contentious, which were required to handle the slightly different structure for the invoice items within the upcoming invoice, and to support things like not re-fetching the data from Stripe and not saving the created object.  Both of the latter could be useful for some additional optimisations later.

I'll be happy to discuss any part of the PR during review!
## Test Output

Ran: `python runtests.py --skip-utc`

```
Welcome to the dj-stripe test suite.

Step 1: Running unit tests.

nosetests . --verbosity=1
Creating test database for alias 'default'...
................................................................................................................................................................................................................S...S.........................................
----------------------------------------------------------------------
Ran 254 tests in 5.540s

OK (SKIP=2)
Destroying test database for alias 'default'...

Step 2: Generating coverage results.

Name                                             Stmts   Miss Branch BrPart  Cover   Missing
--------------------------------------------------------------------------------------------
djstripe/context_managers.py                         8      0      0      0   100%
djstripe/contrib/rest_framework/permissions.py       9      0      0      0   100%
djstripe/contrib/rest_framework/serializers.py      11      0      0      0   100%
djstripe/contrib/rest_framework/urls.py              5      0      0      0   100%
djstripe/contrib/rest_framework/views.py            36      0      2      0   100%
djstripe/decorators.py                              19      0      4      0   100%
djstripe/event_handlers.py                          44      0     16      0   100%
djstripe/exceptions.py                               7      0      0      0   100%
djstripe/fields.py                                  76      0     18      0   100%
djstripe/forms.py                                    7      0      0      0   100%
djstripe/managers.py                                37      0      0      0   100%
djstripe/middleware.py                              36      0     18      0   100%
djstripe/mixins.py                                  26      0      2      0   100%
djstripe/models.py                                 371      0    116      0   100%
djstripe/settings.py                                43      0     12      0   100%
djstripe/signals.py                                  3      0      0      0   100%
djstripe/stripe_objects.py                         485      0     75      0   100%
djstripe/sync.py                                    29      0      4      0   100%
djstripe/templatetags/djstripe_tags.py              20      0      4      0   100%
djstripe/urls.py                                     6      0      0      0   100%
djstripe/utils.py                                   35      0     14      0   100%
djstripe/views.py                                  133      0     22      0   100%
djstripe/webhooks.py                                25      0      8      0   100%
--------------------------------------------------------------------------------------------
TOTAL                                             1471      0    315      0   100%

Step 3: Checking for pep8 errors.

pep8 errors:
----------------------------------------------------------------------
None

Tests completed successfully with no errors. Congrats!
```

Commit:
Add support for upcoming invoices

Also adds migrations (including missing migrations).

kavdev commented on djstripe/models.py:
```diff
@@ -541,20 +542,94 @@ def _attach_objects_hook(self, cls, data):
+        if not self.pk:
+            # InvoiceItems need a saved invoice because they're associated via
+            # a RelatedManager.  It might be better to make this pattern more
+            # generic with some sort-of _attach_objects_post_save_hook().
```
I agree with this. Would you mind adding that post save hook? (Another PR is fine if that would cause any conflicts in the PR chain)

kavdev commented on djstripe/models.py:
```diff
@@ -541,20 +542,94 @@ def _attach_objects_hook(self, cls, data):
+            # generic with some sort-of _attach_objects_post_save_hook().
+            self.save()
+
+        cls._stripe_object_to_invoice_items(InvoiceItem, data, self)
```
Please explicitly state the kwargs here.

kavdev commented on djstripe/models.py:
```diff
@@ -541,20 +542,94 @@ def _attach_objects_hook(self, cls, data):
+
+        In order to provide a consistent view of invoices, it is necessary to
+        examine the invoice items for the associated subscription plan rather
+        than the top-level link.  The reason for this is that the subscription
```
I'm not sure what this second sentence is meant to say.

Also, on a pedantic note, please use one space after a period.

kavdev commented on djstripe/models.py:
```diff
@@ -541,20 +542,94 @@ def _attach_objects_hook(self, cls, data):
+        :returns: The associated plan for the invoice.
+        :rtype: ``djstripe.models.Plan``
+        """
+        for item in self.invoiceitems.all():
```
Please change to `for invoiceitem in ...`

Commit:
Add fixes for PR #320 review

Adds a new _attach_objects_post_save_hook, for attaching Stripe objects
to an instance only after it is saved.

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -178,28 +181,47 @@ def _attach_objects_hook(self, cls, data):
         instance._attach_objects_hook(cls, data)
-        instance.save()
-
+        if save:
```
Wrap these two lines in newlines please

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -178,28 +181,47 @@ def _attach_objects_hook(self, cls, data):
 
     @classmethod
-    def _get_or_create_from_stripe_object(cls, data, field_name="id"):
+    def _get_or_create_from_stripe_object(cls, data, field_name="id",
```
This can stay on one line

lskillen replied:
That pushes it beyond 80 characters - Is that acceptable?  We usually keep ours below it in line with PEP-8 (and it is likely a common theme throughout any submits I've made), but this is your project and I'll follow whatever guidelines you lay out.  Could also bring it down a line instead?

```
    def _get_or_create_from_stripe_object(
             cls, data, field_name="id", refetch=True, save=True):
        field = data.get(field_name)
```

lskillen replied:
OK, seen your further comments later that E501 is ignored - Noted!

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -262,6 +284,54 @@ def _stripe_object_to_invoice(cls, target_cls, data):
+        :param data: The data dictionary received from the Stripe API.
+        :type data: dict
+        """
+        lines = data.get("lines")
```
line break before this

lskillen replied:
Sorry, which line number are you referring to (that you want the line break on)?  Between 300-301?

lskillen replied:
Nevermind, got it based on your other comments that it's between docstring and body. ;-)

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -262,6 +284,54 @@ def _stripe_object_to_invoice(cls, target_cls, data):
+        if not lines:
+            return []
+
+        items = []
```
use `invoiceitems` here for consistency

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -262,6 +284,54 @@ def _stripe_object_to_invoice(cls, target_cls, data):
+        :param target_cls: The target class to instantiate per invoice item.
+        :type target_cls: ``StripeInvoiceItem``
+        :param data: The data dictionary received from the Stripe API.
+        :type data: dict
```
a param/type declaration for the invoice parameter is missing.

lskillen replied:
Good catch, missed that.

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -262,6 +284,54 @@ def _stripe_object_to_invoice(cls, target_cls, data):
+                    if not line["id"].startswith(invoice.stripe_id):
+                        line["id"] = "{invoice_id}-{sub_id}".format(
+                            invoice_id=invoice.stripe_id,
+                            sub_id=line["id"])
```
please use `subscription_id` here

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -262,6 +303,54 @@ def _stripe_object_to_invoice(cls, target_cls, data):
+                save = False
+
+            line.setdefault("customer", invoice.customer.stripe_id)
+            line.setdefault("date", int(invoice.date.strftime("%s")))
```
Is this meant to convert a datetime to a unix timestamp? If so, please use `calendar.timegm(invoice.date.timetuple())` instead.

lskillen replied:
Yep, that's right - Looks like we can borrow from Django instead though: 

```
from django.utils import dateformat
line.setdefault("date", dateformat.format(invoice.date, 'U'))
```

I looked at the implementation and it's pretty similar to your suggestion with the main difference that it uses `utctimetuple` if the datetime is timezone aware:

```
    def U(self):
        "Seconds since the Unix epoch (January 1 1970 00:00:00 GMT)"
        if isinstance(self.data, datetime.datetime) and is_aware(self.data):
            return int(calendar.timegm(self.data.utctimetuple()))
        else:
            return int(time.mktime(self.data.timetuple()))
```

Thoughts?

kavdev replied:
@lskillen: Amazing find! The calendar version is used throughout dj-stripe. I'll refactor after all of these PRs are merged.

kavdev replied:
#328

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -262,6 +303,54 @@ def _stripe_object_to_invoice(cls, target_cls, data):
+            line.setdefault("customer", invoice.customer.stripe_id)
+            line.setdefault("date", int(invoice.date.strftime("%s")))
+
+            item = target_cls._get_or_create_from_stripe_object(
```
Instead of indexing the tuple, please use `item, _created = ...`

lskillen replied:
That works better for me - In this case I was just following existing code for consistency.

One question: Do you mind if I use `_` instead of `_created`, or would you still prefer `_created`?   The reason I ask is that I get warnings for unused variable but `_` is accepted as a dummy variable, usually.

kavdev replied:
@lskillen: Is there a way you can configure your linter to overlook variables starting with `_`?

I believe the convention is indeed just the `_`, but that makes things unclear in case we do end up wanting to use the variable down the road.

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -1097,6 +1187,61 @@ def _stripe_object_to_charge(cls, target_cls, data):
             return target_cls._get_or_create_from_stripe_object(data, "charge")[0]
 
+    @classmethod
+    def upcoming(cls, api_key=settings.STRIPE_SECRET_KEY, **kwargs):
```
Go ahead and specify parameters

kavdev replied:
Also, since this is for a specific customer, add this to Customer as an instance method -- maybe `Customer.upcoming_invoice()`. Follow the convention of returning the stripe object in stripe_objects.py and returning the dj-stripe model instance in models.py.

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -1097,6 +1187,61 @@ def _stripe_object_to_charge(cls, target_cls, data):
+
+        try:
+            data = cls._api().upcoming(api_key=api_key, **kwargs)
+        except InvalidRequestError as ex:
```
Use `exc` instead of `ex`

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -1097,6 +1187,61 @@ def _stripe_object_to_charge(cls, target_cls, data):
+            data = cls._api().upcoming(api_key=api_key, **kwargs)
+        except InvalidRequestError as ex:
+            if str(ex) != "Nothing to invoice for customer":
+                six.reraise(*sys.exc_info())
```
what's the difference between this and just using `raise`?

lskillen replied:
It basically means that the stack trace will continue up into the library rather than stopping at the point of where the exception is raised again manually, which is perhaps "good behaviour" if not translating the exception into a different one. This will only happen where it is supported and probably requires the exception chaining support that python3.x has (as per [PEP-3134](https://www.python.org/dev/peps/pep-3134/)).

Here's what the six documentation says for reraise:

> Reraise an exception, possibly with a different traceback. **In the simple case, `reraise(*sys.exc_info())` with an active exception (in an except block) reraises the current exception with the last traceback.** A different traceback can be specified with the exc_traceback parameter. Note that since the exception reraising is done within the reraise() function, Python will attach the call frame of reraise() to whatever traceback is raised.

kavdev replied:
Alright, another refactor to add to my list :)

kavdev replied:
#328

kavdev commented on tests/test_invoice.py:
```diff
@@ -11,24 +12,29 @@
+        self.user = get_user_model().objects.create_user(
+            username="pydanny", email="pydanny@gmail.com")
+        self.customer = Customer.objects.create(
+            subscriber=self.user, stripe_id=FAKE_CUSTOMER["id"], currency="usd")
```
These can stay on one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -11,24 +12,29 @@
+from djstripe.models import Customer, Invoice, Account, UpcomingInvoice
+from djstripe.models import InvalidRequestError
+
+from tests import (FAKE_INVOICE, FAKE_CHARGE, FAKE_CUSTOMER,
```
These imports are fine on two lines. Note that we exclude PEP8 E501 from our linting.

kavdev commented on tests/test_invoice.py:
```diff
@@ -113,9 +114,10 @@ def test_status_forgiven(self, charge_retrieve_mock, customer_retrieve_mock, sub
-    @patch("stripe.Customer.retrieve", return_value=deepcopy(FAKE_CUSTOMER))
     @patch("stripe.Charge.retrieve", return_value=deepcopy(FAKE_CHARGE))
-    def test_status_closed(self, charge_retrieve_mock, customer_retrieve_mock, subscription_retrieve_mock, default_account_mock):
+    def test_status_closed(self, charge_retrieve_mock,
```
one line here

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
-    @patch("stripe.Customer.retrieve", return_value=deepcopy(FAKE_CUSTOMER))
     @patch("stripe.Charge.retrieve", return_value=deepcopy(FAKE_CHARGE))
-    def test_sync_no_subscription(self, charge_retrieve_mock, customer_retrieve_mock, subscription_retrieve_mock, default_account_mock):
+    def test_sync_no_subscription(self, charge_retrieve_mock,
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+        charge_retrieve_mock.assert_called_once_with(
+            api_key=ANY, expand=ANY, id=FAKE_CHARGE["id"])
+        plan_retrieve_mock.assert_called_once_with(
+            api_key=ANY, expand=ANY, id=FAKE_PLAN["id"])
```
one line for each of these

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+    @patch("djstripe.models.Account.get_default_account")
+    @patch("stripe.Subscription.retrieve", return_value=deepcopy(FAKE_SUBSCRIPTION))
+    @patch("stripe.Charge.retrieve", return_value=deepcopy(FAKE_CHARGE))
+    def test_invoice_with_subscription_invoice_items(
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+        self.assertEquals(1, len(items))
+        item_id = "{invoice_id}-{sub_id}".format(
+            invoice_id=invoice.stripe_id,
+            sub_id=FAKE_SUBSCRIPTION["id"])  # composite key
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+    @patch("djstripe.models.Account.get_default_account")
+    @patch("stripe.Subscription.retrieve", return_value=deepcopy(FAKE_SUBSCRIPTION))
+    @patch("stripe.Charge.retrieve", return_value=deepcopy(FAKE_CHARGE))
+    def test_invoice_with_no_invoice_items(
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+    @patch("djstripe.models.Account.get_default_account")
+    @patch("stripe.Subscription.retrieve", return_value=deepcopy(FAKE_SUBSCRIPTION))
+    @patch("stripe.Charge.retrieve", return_value=deepcopy(FAKE_CHARGE))
+    def test_invoice_with_non_subscription_invoice_items(
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+    @patch("djstripe.models.Account.get_default_account")
+    @patch("stripe.Subscription.retrieve", return_value=deepcopy(FAKE_SUBSCRIPTION))
+    @patch("stripe.Charge.retrieve", return_value=deepcopy(FAKE_CHARGE))
+    def test_invoice_plan_from_invoice_items(
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+    @patch("djstripe.models.Account.get_default_account")
+    @patch("stripe.Subscription.retrieve", return_value=deepcopy(FAKE_SUBSCRIPTION))
+    @patch("stripe.Charge.retrieve", return_value=deepcopy(FAKE_CHARGE))
+    def test_invoice_plan_from_subscription(
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+    @patch("djstripe.models.Account.get_default_account")
+    @patch("stripe.Subscription.retrieve", return_value=deepcopy(FAKE_SUBSCRIPTION))
+    @patch("stripe.Charge.retrieve", return_value=deepcopy(FAKE_CHARGE))
+    def test_invoice_without_plan(
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+    @patch("stripe.Plan.retrieve", return_value=deepcopy(FAKE_PLAN))
+    @patch("stripe.Subscription.retrieve", return_value=deepcopy(FAKE_SUBSCRIPTION))
+    @patch("stripe.Invoice.upcoming", return_value=deepcopy(FAKE_UPCOMING_INVOICE))
+    def test_upcoming_invoice(self, invoice_upcoming_mock,
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+        self.assertIsNone(invoice.save())
+
+        subscription_retrieve_mock.assert_called_once_with(
+            api_key=ANY, expand=ANY, id=FAKE_SUBSCRIPTION["id"])
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+        self.assertIsNotNone(invoice.plan)
+
+    @patch("stripe.Invoice.upcoming", side_effect=InvalidRequestError(
+        "Nothing to invoice for customer", None))
```
one line

kavdev commented on tests/test_invoice.py:
```diff
@@ -140,14 +141,160 @@ def test_sync_send_emails_false(self, charge_retrieve_mock, customer_retrieve_mo
+        self.assertIsNone(invoice)
+
+    @patch("stripe.Invoice.upcoming", side_effect=InvalidRequestError(
+        "Some other error", None))
```
one line

kavdev commented on djstripe/models.py:
```diff
@@ -542,35 +542,29 @@ def _attach_objects_hook(self, cls, data):
-        downloadable invoices).
+        extract it from the first invoice item with a plan, rather than getting
+        it from the subscription by foreign key.  The reason for this is that
+        the subscription plan when requested by the customer, but the invoice
```
Sorry, meant second sentence of this paragraph, but the clarification you've made helps.

The reason for this is the subscription plan ...(what?)... when requested by the customer

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -1097,6 +1190,76 @@ def _stripe_object_to_charge(cls, target_cls, data):
             return target_cls._get_or_create_from_stripe_object(data, "charge")[0]
 
+    @classmethod
+    def upcoming(cls, api_key=settings.STRIPE_SECRET_KEY, customer=None, coupon=None, subscription=None, subscription_plan=None, subscription_prorate=None, subscription_proration_date=None, subscription_quantity=None, subscription_trial_end=None, **kwargs):
```
use a double indent on this second row. If you want, three rows is okay here due to the sheer number of kwargs.

lskillen replied:
Sorry, that was made before the clarification about line length - Fixing. :-)

kavdev commented on djstripe/models.py:
```diff
@@ -541,20 +542,95 @@ def _attach_objects_hook(self, cls, data):
+    def _attach_objects_post_save_hook(self, cls, data):
+        # InvoiceItems need a saved invoice because they're associated via a
+        # RelatedManager, so this must be done as part of the post save hook.
+        cls._stripe_object_to_invoice_items(InvoiceItem, data, self)
```
"Please explicitly state the kwargs here."

lskillen replied:
OK, done this - Actually had misread this earlier and thought it meant something else.

kavdev commented on djstripe/models.py:
```diff
@@ -541,20 +542,95 @@ def _attach_objects_hook(self, cls, data):
+
+        In order to provide a consistent view of invoices, it is necessary to
+        extract it from the first invoice item with a plan, rather than getting
+        it from the subscription by foreign key.
```
"it"? :)

lskillen replied:
A lot of its, isn't ... Trying not to say it. :-)

kavdev replied:
@lskillen: I should clarify: What is the subject in this sentence? What are you extracting? The plan ID?

lskillen replied:
The plan object - It takes the plan object from the first invoice item that has one so perhaps something like:

> In order to provide a consistent view of invoices, the plan object should be taken from the first invoice item that has one, rather than using the plan associated with the subscription.

I'll defer to you for now - I think 3AM is a tad late to be writing good prose.  It'll be a lot easier  tomorrow. ;-)

kavdev replied:
That's perfect. I'll look forward to the final fixes tomorrow.

kavdev commented on djstripe/models.py:
```diff
@@ -541,20 +542,95 @@ def _attach_objects_hook(self, cls, data):
+
+    def _attach_objects_hook(self, cls, data):
+        super(UpcomingInvoice, self)._attach_objects_hook(cls, data)
+        self._items = cls._stripe_object_to_invoice_items(InvoiceItem, data, self)
```
Explicitly state kwargs

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -173,33 +176,71 @@ def _attach_objects_hook(self, cls, data):
-        Create a model instance using the given data object from Stripe
+        Instantiates a model instance using the provided data object received
+        from Stripe, and saves it to the database if specified.
+        :param data: The data dictionary received from the Stripe API.
```
newline before

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -173,33 +176,71 @@ def _attach_objects_hook(self, cls, data):
+        if save:
+            instance.save()
+
+        instance._attach_objects_post_save_hook(cls, data)
```
newline after (before the return statement)

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -262,6 +303,57 @@ def _stripe_object_to_invoice(cls, target_cls, data):
+                save = False
+
+            line.setdefault("customer", invoice.customer.stripe_id)
+            line.setdefault("date", int(dateformat.format(invoice.date, 'U')))
```
I believe the int cast is now redundant

lskillen replied:
I thought so too, but it isn't - It returns a unicode string in my version of Django.  I'll need to look at Django on GitHub to see if that's changed or if there's something else going on.

lskillen replied:
Looks like that's [how it is designed](https://github.com/django/django/blob/31e0314979f6aa1c66828b55dd222c32aad94e9e/django/utils/dateformat.py#L35-L42) unfortunately - Although the function itself does cast to int explicitly, the wrapper for dateformat forces it back to a string.  It's slightly poor form that we go through int -> string -> int though.

kavdev replied:
~~Interesting. I'll open up a ticket with Django and see what happens...~~ Just kidding. They closed their github issue tracker.

kavdev commented on djstripe/models.py:
```diff
@@ -547,6 +557,24 @@ def _attach_objects_post_save_hook(self, cls, data):
+        stripe_invoice = StripeInvoice.upcoming(**kwargs)
+
+        if stripe_invoice:
+            return cls._create_from_stripe_object(stripe_invoice, save=False)
```
Is the class check in 567-571 necessary? You could just use `UpcomingInvoice` on this line instead of `cls`. If that was put in for extendibility, I'd rather go with the class replacement settings idea from a previous conversation.

lskillen replied:
Yes, it was (originally), but I'll remove this and the extensible support can come later.

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -1237,14 +1237,6 @@ def upcoming(cls, api_key=settings.STRIPE_SECRET_KEY, customer=None, coupon=None
-                    "Class argument needs to be derived from UpcomingInvoice")
-
         try:
             data = cls._api().upcoming(
```
`upcoming_stripe_invoice` might be a better variable name

kavdev commented on djstripe/models.py:
```diff
@@ -346,6 +346,16 @@ def add_card(self, source, set_default=True):
+    def upcoming_invoice(self, **kwargs):
+        """ Gets the upcoming preview invoice (singular) for this customer.
+
+        See the documentation in ``djstripe.stripe_objects.StripeInvoice.upcoming()` for details on method parameters.
```
Change this line to `See `StripeInvoice.upcoming() <#djstripe.stripe_objects.StripeInvoice.upcoming>`__`

kavdev commented on djstripe/models.py:
```diff
@@ -346,6 +346,16 @@ def add_card(self, source, set_default=True):
+        """ Gets the upcoming preview invoice (singular) for this customer.
+
+        See the documentation in ``djstripe.stripe_objects.StripeInvoice.upcoming()` for details on method parameters.
+        The ``customer`` parameter is set to the customer object this method is being called on.
```
Instead:

The `customer` argument to the `upcoming()` call is automatically set by this method.

kavdev commented on djstripe/models.py:
```diff
@@ -547,6 +557,24 @@ def _attach_objects_post_save_hook(self, cls, data):
+        """ Gets the upcoming preview invoice (singular) for a customer.
+
+        See the documentation in ``djstripe.stripe_objects.StripeInvoice.upcoming()` for details on method parameters.
+        """
```
Go ahead and remove this docstring, as it will automatically inherit the docstring from StripeInvoice.upcoming()

kavdev commented on docs/models.rst:
```diff
@@ -125,6 +126,8 @@ Invoice
     .. autoattribute:: djstripe.models.Invoice.status
```
`.. autoattribute:: djstripe.models.Invoice.plan`

kavdev commented on docs/models.rst:
```diff
@@ -125,6 +126,8 @@ Invoice
     
     .. automethod:: djstripe.models.Invoice.retry
+
+    .. automethod:: djstripe.models.Invoice.upcoming
```
unnecessarry new line (methods are grouped together)

kavdev commented on djstripe/stripe_objects.py:
```diff
@@ -1097,6 +1192,69 @@ def _stripe_object_to_charge(cls, target_cls, data):
+        :rtype: ``djstripe.models.UpcomingInvoice``
+        """
+        try:
+            data = cls._api().upcoming(
```
Use `upcoming_stripe_invoice` here as well (instead of `data`) and that'll be the last fix
